New function `uid_expunge`, which requires the capability UIDPLUS.

axoroll7 opened the pull request:
The function `expunge` can raise an error at the server side, when the function is utilized with *messages*, and the server does not support the capability UIDPLUS.

Closes #496

mjs commented:
Thanks very much for this. The change looks good but could you address the formatting issue mentioned by black?

mjs commented:
Also, I've just fixed up the project's Github Actions config so if you rebase onto master or merge master the actions should work now.

axoroll7 commented:
Hello,
Thank you for your amazing library !

I updated the pull request, to address the linter issue.
I added some context in the function description.
English is not my first language, I would like you to read carefully each word to correct me.

mjs commented on imapclient/imapclient.py:
```diff
@@ -1505,6 +1505,17 @@ def expunge(self, messages=None):
 
+    @require_capability("UIDPLUS")
+    def uid_expunge(self, messages):
+        """Same functionality as ``expunge``, but *messages* must be
+        specified, and the capability UIDPLUS is tested beforehand. It should
+        be more fail-proof than ``expunge`` with *messages*, which cannot be
+        updated to prevent breaking compability with existing codebase.
```
I don't think the docstring requires quite as much detail. Something like this should do:

>  Expunge deleted messages with the specified message ids from the folder. This requires the UIDPLUS capability.
>
> See :rfc:`4315#section-2.1` section 2.1 for more details.

 Then please extend the docstring on the expunge method to discourage use of `messages` there. Something like:

> Use of the *messages* argument is discouraged. Please see the uid_expunge method instead.

I would be ok with this replacing the existing paragraph about the `messages` argument.

Thanks!

axoroll7 replied:
@mjs 
Your choice of words is better than mine, I am using it without modification. New commit e0cbf56 and ea5ef55

axoroll7 commented:
@mjs 
Your choice of words is better than mine, I am using it without modification. New commit e0cbf56 and ea5ef55

axoroll7 commented:
Depreciation notice, first line ?

mjs reviewed (approved):
This is great. Thanks very much.
